uast/transformer/semantic.go

```diff
@@ -161,96 +164,102 @@ func (c *commentElems) isTab(r rune) bool {
 	return false
 }
 
-func (c *commentElems) Split(text string) bool {
-	if c.DoTrim {
-		text = strings.TrimLeftFunc(text, unicode.IsSpace)
+func max(x, y int) int {
+    if x > y {
+        return x
+    }
+    return y
+}
+
+// Returns the first index i of runes satisfying f(runes[i]), len(runes) otherwise
+func firstIndexFunc(runes []rune, f func(r rune) bool) int {
+	for i, r := range runes {
+		if (f(r)) {
+			return i;
+		}
 	}
+	return len(runes);
+}
 
-	if !strings.HasPrefix(text, c.StartToken) || !strings.HasSuffix(text, c.EndToken) {
-		return false
+// Returns the last index i of runes satisfying f(runes[i]), -1 otherwise
+func lastIndexFunc(runes []rune, f func(r rune) bool) int {
+	for i := range runes {
+		j := len(runes) - 1 - i
+		if (f(runes[j])) {
+			return j;
+		}
 	}
-	text = strings.TrimPrefix(text, c.StartToken)
-	text = strings.TrimSuffix(text, c.EndToken)
+	return -1;
+}
 
-	// find prefix (if not already trimmed)
-	i := strings.IndexFunc(text, func(r rune) bool {
-		return !c.isTab(r)
-	})
+func (c *commentElems) findPrefix(f func(r rune) bool) {
+	runes := []rune(c.Text)
+	i := firstIndexFunc(runes, f)
+	c.Prefix = string(runes[:i])
+	c.Text = string(runes[i:])
+}
 
-	c.Prefix = ""
+func (c *commentElems) findSuffix(f func(r rune) bool) {
+	runes := []rune(c.Text)
+	i := lastIndexFunc(runes, f) + 1
+	c.Suffix = string(runes[i:])
+	c.Text = string(runes[:i])
+}
 
-	if i >= 0 {
-		c.Prefix = text[:i]
-		text = text[i:]
-	} else {
-		c.Prefix = text
-		text = ""
+func commonPrefix(a []rune, b []rune) []rune {
+	r := []rune{}
+	for i := 0; i < len(a) && i < len(b); i++ {
+		if (a[i] != b[i]) {
+			break;
+		} else {
+			r = append(r, a[i])
```

Comment:
do we need to create a new slice? Cannot we just return `a[:i+1]`

Reply:
Changed to avoid creation of extra slice!